### PR TITLE
Add OpenTelemetry-based tracing to the backend

### PR DESCRIPTION
**Pull request by aapeliv**

### CODE REVIEW

**Comment by aapeliv** [comment]
@spreeni would you be able to have a look at this please?

**Comment by aapeliv** [comment]
Screenshot from 2024-08-28 10-53-29

Generates this kind of data

File: app/docker-compose.yml [hunk 1]
```diff
@@ -42,9 +41,19 @@ services:
...
     ports:
       - 1025:1025
       - 1080:1080
+
+  jaeger:
+    image: jaegertracing/all-in-one:1.60
+    ports:
+      - "16686:16686"
+    expose:
+      - 4317
```
Comment (spreeni) [inline app/docker-compose.yml:57]: Ah nice, never used `expose` so far, but neat

File: app/backend/src/couchers/tracing.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        SQLAlchemyInstrumentor().instrument(engine=_get_base_engine(), enable_commenter=True, commenter_options={})
+        SQLAlchemyInstrumentor().instrument(
+            engine=_get_worker_base_engine(), enable_commenter=True, commenter_options={}
+        )
+
+        trace.set_tracer_provider(TracerProvider(resource=Resource(attributes={"service.name": "backend"})))
+
+        trace.get_tracer_provider().add_span_processor(
+            BatchSpanProcessor(OTLPSpanExporter(endpoint=config["OPENTELEMETRY_ENDPOINT"], insecure=True))
+        )
```
Comment (spreeni) [inline app/backend/src/couchers/tracing.py:29]: This is exciting! Judging by https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/grpc/grpc.html, the TracerProvider should be set up before the Instrumentors logic-wise, no?

File: app/backend/src/couchers/interceptors.py [hunk 3]
```diff
@@ -178,6 +181,47 @@ def intercept_service(self, continuation, handler_call_details):
...
+                rollspan.set_attribute("rpc.service", service_name)
+                rollspan.set_attribute("rpc.method", method_name)
+
+                rollspan.set_attribute("rpc.thread", get_ident())
+                rollspan.set_attribute("rpc.pid", getpid())
+
+                res = prev_func(request, context)
+
+                rollspan.set_attribute("web.user_agent", headers.get("user-agent") or "")
+                rollspan.set_attribute("web.ip_address", headers.get("x-couchers-real-ip") or "")
```
Comment (spreeni) [inline app/backend/src/couchers/interceptors.py:214]: Is this always sent from the Frontend?

**Review by spreeni — COMMENTED** [review]
Looks good to me, sorry for missing this! Did we already get any insights into relevant bottlenecks in the code?

Very cool to have all the Instrumentors ready, might look at it for work as well. :)

Have you also considered exporting OpenTelemetry data simply to Sentry? Not sure if that is something desirable though.

**Pull request merged**